pkg/types/aws_resource.go

```diff
@@ -51,4 +53,8 @@ type AWSResource interface {
 	// apimachinery/runtime.Object and the Kubernetes
 	// apimachinery/apis/meta/v1.Object interfaces
 	RuntimeMetaObject() RuntimeMetaObject
+	// SetObjectMeta sets the ObjectMeta field for the resource
+	SetObjectMeta(meta metav1.ObjectMeta)
+	// SetARN will set the ARN field in the resource's metadata status field
+	SetARN(*ackv1alpha1.AWSResourceName)
```

Comment:
What we need is an interface that will help the `AWSResourceManager` determine which field in the `acktypes.AWSResource`'s underlyihng `Spec` or `Status` struct maps to each field in a `ackcorev1alpha1.AWSIdentifiers` struct. We already know that the `ackcorev1alpha1.AWSIdentifiers.ARN` field maps to the `Status.ACKMetadata.ARN` field because we standardize on that. What we don't really know is which field in the `Spec` corresponds to the `ackcorev1alpha1.AWSIdentifiers.Name` and `ackcorev1alpha1.AWSIdentifiers.ID` field.

Something like this might work?

```go
// AWSIdentifierFieldMapper helps a resource manager determine which fields
// in a resource represent name or generated IDs.
type AWSIdentifierFieldMapper interface {
    // NameField returns the path to the field in the underlying resource's `Spec`
    // or `Status` struct that represents the resource's name. This will be empty
    // if there is no name field for the resource.
    NameField() string
    // IDField returns the path to the field in the underlying resource's `Spec`
    // or `Status` struct that represents the resource's AWS-generated ID field.
    // This will be empty if there is no such field for the resource.
    IDField() string
}
```

We could then have the `AWSResourceDescriptor` interface have a method that returns a `AWSIdentifierFieldMapper`-implementing struct:

```go
type AWSResourceDescriptor interface {
// ...
    // IdentifierFieldMapper returns an `AWSIdentifierFieldMapper` that instructs callers
    // which fields in the `AWSResource` represent name or generated IDs.
    IdentifierFieldMapper() AWSIdentifierFieldMapper
}
```

Furthemore, instead of a `SetNameField` and `SetARN` method (and I presume you would eventually want a `SetID` method?) on the `AWSResource` interface above, how about just adding a `SetIdentifiers` method that accepts a single parameter of type `ackcorev1alpha1.AWSIdentifiers`? That way, the `AdoptedResourceReconciler` can construct a new `AWSResource` and set its identifier fields using a single method.

So, instead of this code that you have above in the `AdoptedResourceReconciler`:

```go
	// Set spec fields prior to reading
	if desired.Spec.AWS.Name != nil {
		readableResource.SetNameField(*desired.Spec.AWS.Name)
	} else if desired.Spec.AWS.ID != nil {
		readableResource.SetNameField(*desired.Spec.AWS.ID)
	} else if desired.Spec.AWS.ARN != nil {
		readableResource.SetARN(desired.Spec.AWS.ARN)
	} else {
		return fmt.Errorf("must provide at least one value for identifier")
	}
```

you could just do this instead:

```go
    readableResource.SetIdentifiers(*desired.Spec.AWS)
```

and let the plumbing inside the controller's `pkg/resource/$RESOURCE/descriptor.go` and `pkg/resource/$RESOURCE/resource.go` files use the `AWSIdentiferFieldMapper` to determine the right fields in the underlying resource to pull out.